### PR TITLE
Lite: Concatenation Op Refactored

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
1:> Code refactored for reference_ops usage as both optimized and reference are same.
2:> Error message corrected.
3:> Wrong restriction removed.
4:> Test case added for Quantizied Input.

### CODE REVIEW

**Comment by jdduke** [comment]
Even though both are the same now, when we add an optimized version in optimized_ops.h, we'll have a way of selecting between ref/opt using a custom resolver, so I would leave that code as it is.

**Comment by ANSHUMAN87** [comment]
> Even though both are the same now, when we add an optimized version in optimized_ops.h, we'll have a way of selecting between ref/opt using a custom resolver, so I would leave that code as it is.

@jdduke : Completely agree with you. But i think, the code is confusing, even i had to spend some time to find out, that both are same. Also it is not very big changes to accommodate optimization kernel if added in future. May be what we can do, we can keep the Enum(KernelType), and mark a TODO by optimization type, to be added in future. What do you think?

**Comment by srjoglekar246** [comment]
This is a convention we follow in most op-code, so it won't make sense to modify one implementation. IMO we can keep the current logic as-is.

**Comment by ANSHUMAN87** [comment]
@srjoglekar246: I found many op-code which follow convention as I mentioned in my previous comment. I think if we keep the current code it loses the readability also. It's better we follow the convention which is more readable.

File: tensorflow/lite/kernels/concatenation.cc [hunk 1]
```diff
@@ -54,7 +48,6 @@ TfLiteStatus Prepare(TfLiteContext* context, TfLiteNode* node) {
 
   // TODO(ahentz): These are limitations of our implementation that could be
   // removed with a bit of effort.
-  TF_LITE_ENSURE(context, t0->dims->size <= 4);
```
Comment (jdduke) [inline tensorflow/lite/kernels/concatenation.cc]: Please avoid coupling refactorings with behavioral changes. This should be its own CL, with its own test, separate from any kind of refactoring.

**Review by jdduke — CHANGES_REQUESTED** [review]

**Comment by ANSHUMAN87** [comment]
@jdduke: Would you please help conclude this PR, TIA!

**Comment by ANSHUMAN87** [comment]
@jdduke : gentle reminder for conclusion, TIA!

**Comment by jdduke** [comment]
The new tests are fine, but I don't think it's worth removing the kOptimized codepath yet. Instead it'd be good to figure out how to refactor the code to make it easier to have a parallel optimized/reference codepath without the if/else conditions.

**Comment by ANSHUMAN87** [comment]
> The new tests are fine, but I don't think it's worth removing the kOptimized codepath yet. Instead it'd be good to figure out how to refactor the code to make it easier to have a parallel optimized/reference codepath without the if/else conditions.

@jdduke : Thanks for the suggestion! I have handled your comment, please check and approve, Thanks!

File: tensorflow/lite/kernels/concatenation.cc [hunk 2]
```diff
@@ -187,8 +179,6 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
   return kTfLiteOk;
 }
 
-#undef TF_LITE_MACRO_DISPATCH
```
Comment (jdduke) [inline tensorflow/lite/kernels/concatenation.cc]: It's good practice to under the macros.

File: tensorflow/lite/kernels/concatenation.cc [hunk 3]
```diff
@@ -112,72 +112,64 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
...
-      }
+      TF_LITE_CONCATENATION(int32);
       break;
     case kTfLiteUInt8:
-      if (kernel_type == kReference) {
-        TF_LITE_CONCATENATION_QUANTIZED(reference_ops);
-      } else {
-        TF_LITE_CONCATENATION_QUANTIZED(optimized_ops);
-      }
+      TF_LITE_CONCATENATION_QUANTIZED;
```
Comment (jdduke) [inline tensorflow/lite/kernels/concatenation.cc]: Let's make this define a function style invoke with no args

TF_LITE_CONCATENATION_QUANTIZED()

**Comment by ANSHUMAN87** [comment]
@jdduke : Your comments are handled now, please check, Thanks!

**Review by jdduke — APPROVED** [review]

**Pull request merged**